feat: adds api layer to handle email notification

as1729 opened the pull request:
### Ticket #19 

### Description
This PR adds:
1. Ability to get/set email subscription management preferences from the database
2. Adds test coverage and database uniqueness constraints to ensure there are no duplicate subscription preferences for user/agency/notification-type.
3. Adds an API endpoint to get/put these values

### Screenshots / Demo Video

#### Database Changes
<img width="1029" alt="image" src="https://user-images.githubusercontent.com/6497366/212164591-5a381bb3-ec66-414d-a6fd-3b4cff74e956.png">

#### Subscription form when user has permissions to manage the subscription settings.
<img width="504" alt="image" src="https://user-images.githubusercontent.com/6497366/212165163-da0070c7-5fe3-4f49-9189-abc5ad6d13aa.png">


#### Subscription form when user does not have permissions to manage the subscription settings.
<img width="513" alt="image" src="https://user-images.githubusercontent.com/6497366/212165308-86f58cfa-3158-47d8-8f8f-7a537d5de24d.png">


### Testing
- Unit tests

### Checklist
- [x] Provided ticket and description
- [x] Your PR title contains the ISSUE ticket number e.g "86: ..."
- [x] Provided screenshots/demo
- [x] Provided testing information
- [x] Provided adequate test coverage for all new code
- [x] Run automated tests (docker compose exec app yarn test)
- [x] Added PR reviewers
- [ ] Ensure at least 1 review before merging

chrisxkeith commented on packages/server/src/routes/agencies.js:
```diff
@@ -53,6 +54,28 @@ router.delete('/del/:agency', requireAdminUser, async (req, res) => {
+        await setAgencyEmailSubscriptionPreference(agency, preferences);
+        res.sendStatus(200);
+    } catch (e) {
+        console.error(`Unable to update agency email preferences for agency: ${agency} preferences: ${preferences}`);
```
To help developers diagnose errors, I like to include the error (e.g.` e.toString()` I think) as part of the console log. I also like to include it in the page returned to the user. I realize this may be confusing to the end user, but it does help diagnose the problem.

as1729 replied:
good idea! I'll add that in 👍

as1729 replied:
Done 👍

chrisxkeith commented on packages/server/__tests__/db/db.test.js:
```diff
@@ -247,4 +248,107 @@ describe('db', () => {
+            const test = async () => {
+                try {
+                    await db.setAgencyEmailSubscriptionPreference(fixtures.agencies.accountancy.id, {
+                        UNKNOWN: emailConstants.emailSubscriptionStatus.subscribed,
```
Where is UNKNOWN defined? I can't find it anywhere. :(

as1729 replied:
It isn't defined anywhere, I just made it up in this line to make sure that when we pass an unrecognized/unknown value to the function it will raise an error.

chrisxkeith commented:
So does Javascript somehow 'promote' the token UNKNOWN to the string
'UNKNOWN'? As a "Cranky Old Guy Who Prefers Strong Typing" :) I find this
confusing.

Is this some Javascript paradigm/convention that's widely used?

I'm not asking you to change it, I'm just trying to understand how it works.

Thanks!

On Fri, Jan 6, 2023 at 8:22 AM aditya ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In packages/server/__tests__/db/db.test.js
> <https://github.com/usdigitalresponse/usdr-gost/pull/769#discussion_r1063580559>
> :
>
> > +            );
> +            const result = await db.getAgencyEmailSubscriptionPreference(fixtures.agencies.accountancy.id);
> +            expect(result).to.have.all.keys(...Object.values(emailConstants.notificationType));
> +            expect(result[emailConstants.notificationType.grantAssignment]).to.equal(emailConstants.emailSubscriptionStatus.subscribed);
> +            const resultSet = new Set(Object.values(result));
> +            expect(resultSet.size).to.equal(2);
> +            expect(resultSet.has(emailConstants.emailSubscriptionStatus.unsubscribed)).to.equal(true);
> +            expect(resultSet.has(emailConstants.emailSubscriptionStatus.subscribed)).to.equal(true);
> +        });
> +        it('raises error if unknown notification type is supplied', async () => {
> +            let expectedError = '';
> +
> +            const test = async () => {
> +                try {
> +                    await db.setAgencyEmailSubscriptionPreference(fixtures.agencies.accountancy.id, {
> +                        UNKNOWN: emailConstants.emailSubscriptionStatus.subscribed,
>
> It isn't defined anywhere, I just made it up in this line to make sure
> that when we pass an unrecognized/unknown value to the function it will
> raise an error.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/usdigitalresponse/usdr-gost/pull/769#discussion_r1063580559>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA55MVQBEKXVDBUSWLDJPLLWRBBFBANCNFSM6AAAAAATRIDBKA>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>


-- 
-CK
https://github.com/chrisxkeith

as1729 commented:
> So does Javascript somehow 'promote' the token UNKNOWN to the string
'UNKNOWN'? As a "Cranky Old Guy Who Prefers Strong Typing" :) I find this
confusing.

I share the confusion here, but I believe this is how it works. Take a look at this sample code and output:
```js
const obj1 = { foo: 'FOO', bar: 'BAR' };

console.log(obj1);
// { foo: 'FOO', bar: 'BAR' }

console.log(Object.keys(obj1));
// [ 'foo', 'bar' ]

console.log(typeof Object.keys(obj1)[0])
// string
```


> Is this some Javascript paradigm/convention that's widely used?

I'm not sure if its widely used, but our linter prefers that we do not explicitly cast it as a string. For example doing `const obj1 = { 'foo': 'FOO', 'bar': 'BAR' };` will raise an error.

chrisxkeith commented:
OK, I get it now.

const obj1 = { foo: 'FOO', bar: 'BAR' };
console.log(obj1.foo);
FOO
console.log(obj1['foo']);
FOO

Would you like me to accept that Pull Request as is, or wait until you make
that error reporting change?

Thanks!

On Fri, Jan 6, 2023 at 9:11 AM aditya ***@***.***> wrote:

> So does Javascript somehow 'promote' the token UNKNOWN to the string
> 'UNKNOWN'? As a "Cranky Old Guy Who Prefers Strong Typing" :) I find this
> confusing.
>
> I share the confusion here, but I believe this is how it works. Take a
> look at this sample code and output:
>
> const obj1 = { foo: 'FOO', bar: 'BAR' };
> console.log(obj1);// { foo: 'FOO', bar: 'BAR' }
> console.log(Object.keys(obj1));// [ 'foo', 'bar' ]
> console.log(typeof Object.keys(obj1)[0])// string
>
> Is this some Javascript paradigm/convention that's widely used?
>
> I'm not sure if its widely used, but our linter prefers that we do not
> explicitly cast it as a string. For example doing const obj1 = { 'foo':
> 'FOO', 'bar': 'BAR' }; will raise an error.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/usdigitalresponse/usdr-gost/pull/769#issuecomment-1373906813>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA55MVX6GBU4P4KA7EYVW43WRBG27ANCNFSM6AAAAAATRIDBKA>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>


-- 
-CK
https://github.com/chrisxkeith

as1729 commented:
@chrisxkeith apologies for the large review but here's what I changed
- Updated the API endpoints to be on the User-level rather than the Agency-level (This was a product decision).
- Added some front-end code to manage subscription settings in the the ProfileSettings Modal. ( Added screenshots to the PR).
